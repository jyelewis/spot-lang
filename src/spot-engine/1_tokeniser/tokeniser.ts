import { spotKeywords, spotSymbols, SpotToken, TOKEN_TYPES } from './SpotToken';
import { StringReader } from './StringReader';
import assert from 'node:assert';

export function tokeniseCode(code: string): SpotToken[] {
  const tokeniser = new Tokeniser(code);
  return tokeniser.tokenise();
}

class Tokeniser {
  private codeReader: StringReader;
  private tokens: SpotToken[] = [];

  constructor(code: string) {
    this.codeReader = new StringReader(code);
  }

  tokenise(): SpotToken[] {
    while (!this.codeReader.isEOF) {
      if (this.tryMatchComment()) continue;
      if (this.tryMatchKeyword()) continue;
      if (this.tryMatchIdentifier()) continue;
      if (this.tryMatchSymbol()) continue;
      if (this.tryMatchStringTemplate()) continue;
      if (this.tryMatchNumber()) continue;

      this.throwUnrecognizedTokenError();
    }
    return this.tokens;
  }

  private tryMatchComment(): boolean {
    if (!this.codeReader.peak('//')) return false;

    const commentLocation = this.codeReader.cloneLocation();
    this.codeReader.consume('//');
    const commentText = this.codeReader.restOfCurrentLine;
    this.codeReader.consume(commentText);

    this.tokens.push({
      type: TOKEN_TYPES.COMMENT,
      location: commentLocation,
      comment: commentText,
    });
    return true;
  }

  private tryMatchKeyword(): boolean {
    for (const keyword of spotKeywords) {
      if (this.codeReader.peak(keyword)) {
        const location = this.codeReader.cloneLocation();
        this.codeReader.consume(keyword);
        this.tokens.push({
          type: TOKEN_TYPES.KEYWORD,
          location,
          keyword,
        });
        return true;
      }
    }
    return false;
  }

  private tryMatchIdentifier(): boolean {
    // Fixed regex: should not match identifiers starting with numbers
    const identifierMatch = this.codeReader.restOfCurrentLine.match(/^([A-Za-z_][A-Za-z0-9_]*)/);
    if (!identifierMatch) return false;

    const identifier = identifierMatch[0];

    // Ensure identifier isn't a keyword that was missed
    if ((spotKeywords as readonly string[]).includes(identifier)) {
      return false; // Let keyword matcher handle it
    }

    const location = this.codeReader.cloneLocation();
    this.codeReader.consume(identifier);
    this.tokens.push({
      type: TOKEN_TYPES.IDENTIFIER,
      location,
      identifier,
    });
    return true;
  }

  private tryMatchSymbol(): boolean {
    for (const symbol of spotSymbols) {
      if (this.codeReader.peak(symbol)) {
        const location = this.codeReader.cloneLocation();
        this.codeReader.consume(symbol);
        this.tokens.push({
          type: TOKEN_TYPES.SYMBOL,
          location,
          symbol,
        });
        return true;
      }
    }
    return false;
  }

  private tryMatchStringTemplate(): boolean {
    if (!this.codeReader.peak('"')) return false;

    const startLocation = this.codeReader.cloneLocation();

    // Emit string template start token
    this.tokens.push({
      type: TOKEN_TYPES.STRING_TEMPLATE_START,
      location: startLocation,
    });

    this.codeReader.moveAheadBy(1); // consume opening quote

    let literal = '';

    while (!this.codeReader.isEOF && !this.codeReader.peak('"')) {
      const char = this.codeReader.restOfCurrentLine.charAt(0);

      if (char === '\\') {
        // Handle escape sequences
        this.codeReader.moveAheadBy(1);
        if (this.codeReader.isEOF) {
          throw new Error(
            `Unterminated string template at line ${startLocation.line}, column ${startLocation.column}`
          );
        }
        const escapedChar = this.codeReader.restOfCurrentLine.charAt(0);
        literal += this.processEscapeSequence(escapedChar);
        this.codeReader.moveAheadBy(1);
      } else if (char === '{') {
        // Found start of expression - emit current literal if any
        if (literal.length > 0) {
          this.tokens.push({
            type: TOKEN_TYPES.STRING_LITERAL,
            location: this.codeReader.cloneLocation(),
            literal,
          });
          literal = '';
        }

        // Emit expression start token
        const exprStartLocation = this.codeReader.cloneLocation();
        this.tokens.push({
          type: TOKEN_TYPES.STRING_TEMPLATE_EXPRESSION_START,
          location: exprStartLocation,
        });

        this.codeReader.moveAheadBy(1); // consume '{'
        this.codeReader.consumeWhitespace(); // consume any whitespace after '{'

        // Parse the expression inside {}
        this.parseStringTemplateExpression();

        // Consume '}' and emit expression end token
        if (!this.codeReader.peak('}')) {
          throw new Error(
            `Expected '}' to close expression in string template at line ${exprStartLocation.line}, column ${exprStartLocation.column}`
          );
        }

        const exprEndLocation = this.codeReader.cloneLocation();
        this.tokens.push({
          type: TOKEN_TYPES.STRING_TEMPLATE_EXPRESSION_END,
          location: exprEndLocation,
        });

        this.codeReader.moveAheadBy(1); // consume '}'
      } else {
        literal += char;
        this.codeReader.moveAheadBy(1);
      }
    }

    if (this.codeReader.isEOF) {
      throw new Error(
        `Unterminated string template at line ${startLocation.line}, column ${startLocation.column}`
      );
    }

    // Emit final literal if any
    if (literal.length > 0) {
      this.tokens.push({
        type: TOKEN_TYPES.STRING_LITERAL,
        location: this.codeReader.cloneLocation(),
        literal,
      });
    }

    this.codeReader.moveAheadBy(1); // consume closing quote
    this.codeReader.consumeWhitespace(); // consume trailing whitespace

    // Emit string template end token
    this.tokens.push({
      type: TOKEN_TYPES.STRING_TEMPLATE_END,
      location: this.codeReader.cloneLocation(),
    });

    return true;
  }

  private parseStringTemplateExpression(): void {
    // Parse tokens inside the expression until we hit '}'
    let braceDepth = 0;

    while (!this.codeReader.isEOF) {
      // Check if we've hit the closing brace at depth 0
      if (this.codeReader.peak('}') && braceDepth === 0) {
        break;
      }

      // Track nested braces
      if (this.codeReader.peak('{')) {
        braceDepth++;
      } else if (this.codeReader.peak('}')) {
        braceDepth--;
      }

      // Parse individual tokens within the expression
      if (this.tryMatchComment()) continue;
      if (this.tryMatchKeyword()) continue;
      if (this.tryMatchIdentifier()) continue;
      if (this.tryMatchSymbol()) continue;
      if (this.tryMatchStringTemplate()) continue; // Allow nested string templates
      if (this.tryMatchNumber()) continue;

      // If we get here, we have an unrecognized character
      this.throwUnrecognizedTokenError();
    }
  }

  private processEscapeSequence(char: string): string {
    switch (char) {
      case 'n':
        return '\n';
      case 't':
        return '\t';
      case 'r':
        return '\r';
      case '\\':
        return '\\';
      case '"':
        return '"';
      case '0':
        return '\0';
      default:
        // For unknown escape sequences, just return the character as-is
        return char;
    }
  }

  private tryMatchNumber(): boolean {
    const numberMatch = this.codeReader.restOfCurrentLine.match(/^(\d+(?:\.\d+)?)/);
    if (!numberMatch) return false;

    const location = this.codeReader.cloneLocation();
    const numberString = numberMatch[0];
    const isFloat = numberString.includes('.');

    this.codeReader.consume(numberString);

    if (isFloat) {
      const floatValue = parseFloat(numberString);
      assert(!isNaN(floatValue), 'Should have parsed a float');
      this.tokens.push({
        type: TOKEN_TYPES.FLOAT,
        location,
        float: floatValue,
      });
    } else {
      const intValue = parseInt(numberString, 10);
      assert(!isNaN(intValue), 'Should have parsed an integer');
      this.tokens.push({
        type: TOKEN_TYPES.INT,
        location,
        int: intValue,
      });
    }

    return true;
  }

  private throwUnrecognizedTokenError(): never {
    const location = this.codeReader.cloneLocation();
    const context = this.codeReader.restOfCurrentLine.slice(0, 20);
    const char = this.codeReader.restOfCurrentLine.charAt(0);

    throw new Error(
      `Unexpected character '${char}' at line ${location.line}, column ${location.column}\n` +
        `Context: "${context}${context.length === 20 ? '...' : ''}"`
    );
  }
}
