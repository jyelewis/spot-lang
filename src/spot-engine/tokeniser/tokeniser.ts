import { spotKeywords, spotSymbols, SpotToken } from './SpotToken';
import { CodeFileReader } from './CodeFileReader';
import assert from 'node:assert';

export function tokeniseCode(code: string): SpotToken[] {
  const codeReader = new CodeFileReader(code);
  const tokens: SpotToken[] = [];
  while (!codeReader.isEOF) {
    // try to match comments
    if (codeReader.peak('//')) {
      const commentLocation = codeReader.cloneLocation();

      // rest of the line is a comment
      codeReader.consume('//');
      const commentText = codeReader.restOfCurrentLine;
      codeReader.consume(commentText);
      tokens.push({
        type: 'comment',
        location: commentLocation,
        comment: commentText,
      });
    }

    // try to match keyword tokens
    let foundKeyword = false;
    for (const keyword of spotKeywords) {
      if (codeReader.consume(keyword)) {
        foundKeyword = true;
        tokens.push({
          type: 'keyword',
          location: codeReader.cloneLocation(),
          keyword,
        });
        break;
      }
    }

    if (foundKeyword) {
      continue;
    }

    // try to match identifier tokens
    const identifierMatch = codeReader.restOfCurrentLine.match(/^([A-Za-z_][A-Z-a-z0-9_]+)/);
    if (identifierMatch) {
      const identifier = identifierMatch[0];
      codeReader.consume(identifier);
      tokens.push({
        type: 'identifier',
        location: codeReader.cloneLocation(),
        identifier,
      });
      continue;
    }

    let foundSymbol = false;
    for (const symbol of spotSymbols) {
      if (codeReader.consume(symbol)) {
        tokens.push({
          type: 'symbol',
          location: codeReader.cloneLocation(),
          symbol,
        });
        foundSymbol = true;
        break;
      }
    }

    if (foundSymbol) {
      continue;
    }

    // try to read a string literal
    if (codeReader.peak('"')) {
      const startLocation = codeReader.cloneLocation();

      // TODO: revisit with more robust string handling
      const endColumn = codeReader.restOfCurrentLine.substring(1).indexOf('"');

      const literal = codeReader.restOfCurrentLine.substring(1, endColumn + 1);
      const didConsume = codeReader.consume(`"${literal}"`);
      assert(didConsume, 'Should have consumed the string literal');

      tokens.push({
        type: 'string_literal',
        location: startLocation,
        literal,
      });
      continue;
    }

    // try to match a float literal
    const floatMatch = codeReader.restOfCurrentLine.match(/^([0-9]+\.[0-9]+)/);
    if (floatMatch !== null) {
      const floatLocation = codeReader.cloneLocation();
      const floatString = floatMatch[0];
      const floatValue = parseFloat(floatString);
      assert(!isNaN(floatValue), 'Should have parsed a float');

      const didConsume = codeReader.consume(floatString);
      assert(didConsume, 'Should have consumed the float literal');

      tokens.push({
        type: 'float',
        location: floatLocation,
        float: floatValue,
      });
      continue;
    }

    // try to read an integer literal
    const intMatch = codeReader.restOfCurrentLine.match(/^([0-9]+)/);
    if (intMatch !== null) {
      const intLocation = codeReader.cloneLocation();
      const intString = intMatch[0];
      const intValue = parseInt(intString, 10);
      assert(!isNaN(intValue), 'Should have parsed an integer');

      const didConsume = codeReader.consume(intString);
      assert(didConsume, 'Should have consumed the integer literal');

      tokens.push({
        type: 'int',
        location: intLocation,
        int: intValue,
      });
      continue;
    }

    throw new Error(
      'Unrecognised token at ' +
        JSON.stringify(codeReader.cloneLocation()) +
        ': "' +
        codeReader.restOfCurrentLine.slice(0, 10) +
        '"' +
        (codeReader.restOfCurrentLine.length > 10 ? '...' : '')
    );
  }

  return tokens;
}
