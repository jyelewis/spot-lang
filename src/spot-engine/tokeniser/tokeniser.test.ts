import { tokeniseCode } from './tokeniser';

describe('Tokeniser', () => {
  it('Reads hello world example', () => {
    const code = `
    
func main() {
  print("Hello, World!")
}
`;
    const tokens = tokeniseCode(code);
    expect(tokens).toMatchSnapshot();
  });

  it('Variable definitions', () => {
    const code = `
    
func main() {
  let name = "John Doe"
  print("Hello, World!")
}
`;
    const tokens = tokeniseCode(code);
    expect(tokens).toMatchSnapshot();
  });

  it('Comments', () => {
    const code = `
    
func main() {
  // this is a comment
}
`;
    const tokens = tokeniseCode(code);
    expect(
      tokens.find((x) => x.type === 'comment' && x.comment === 'this is a comment')
    ).toBeDefined();
    expect(tokens).toMatchSnapshot();
  });

  it('Integers', () => {
    const code = `
func main() {
  let num = 123
}
`;
    const tokens = tokeniseCode(code);
    expect(tokens.find((x) => x.type === 'int' && x.int === 123)).toBeDefined();
    expect(tokens).toMatchSnapshot();
  });

  it('Floats', () => {
    const code = `
func main() {
  let num = 123.456
}
`;
    const tokens = tokeniseCode(code);
    expect(tokens.find((x) => x.type === 'float' && x.float === 123.456)).toBeDefined();
    expect(tokens).toMatchSnapshot();
  });
});
