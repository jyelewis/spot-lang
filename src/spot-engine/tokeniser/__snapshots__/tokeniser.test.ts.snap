// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokeniser Comments 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 6,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 10,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 11,
      "line": 3,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 3,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 3,
      "line": 4,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "comment": "this is a comment",
    "location": {
      "column": 3,
      "line": 4,
    },
    "type": "comment",
  },
  {
    "location": {
      "column": 1,
      "line": 6,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Floats 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 10,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 3,
      "line": 3,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 7,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "num",
    "location": {
      "column": 11,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 13,
      "line": 3,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "float": 123.456,
    "location": {
      "column": 13,
      "line": 3,
    },
    "type": "float",
  },
  {
    "location": {
      "column": 1,
      "line": 5,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Integers 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 6,
      "line": 2,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 10,
      "line": 2,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 11,
      "line": 2,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 2,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 3,
      "line": 3,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 7,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "num",
    "location": {
      "column": 11,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 13,
      "line": 3,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "int": 123,
    "location": {
      "column": 13,
      "line": 3,
    },
    "type": "int",
  },
  {
    "location": {
      "column": 1,
      "line": 5,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Reads hello world example 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 6,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 10,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 11,
      "line": 3,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 3,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 3,
      "line": 4,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "identifier": "print",
    "location": {
      "column": 8,
      "line": 4,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 9,
      "line": 4,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "Hello, World!",
    "location": {
      "column": 9,
      "line": 4,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 1,
      "line": 5,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 1,
      "line": 6,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;

exports[`Tokeniser Variable definitions 1`] = `
[
  {
    "keyword": "func",
    "location": {
      "column": 6,
      "line": 3,
    },
    "type": "keyword",
  },
  {
    "identifier": "main",
    "location": {
      "column": 10,
      "line": 3,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 11,
      "line": 3,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "location": {
      "column": 13,
      "line": 3,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 3,
      "line": 4,
    },
    "symbol": "{",
    "type": "symbol",
  },
  {
    "keyword": "let",
    "location": {
      "column": 7,
      "line": 4,
    },
    "type": "keyword",
  },
  {
    "identifier": "name",
    "location": {
      "column": 12,
      "line": 4,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 14,
      "line": 4,
    },
    "symbol": "=",
    "type": "symbol",
  },
  {
    "literal": "John Doe",
    "location": {
      "column": 14,
      "line": 4,
    },
    "type": "string_literal",
  },
  {
    "identifier": "print",
    "location": {
      "column": 8,
      "line": 5,
    },
    "type": "identifier",
  },
  {
    "location": {
      "column": 9,
      "line": 5,
    },
    "symbol": "(",
    "type": "symbol",
  },
  {
    "literal": "Hello, World!",
    "location": {
      "column": 9,
      "line": 5,
    },
    "type": "string_literal",
  },
  {
    "location": {
      "column": 1,
      "line": 6,
    },
    "symbol": ")",
    "type": "symbol",
  },
  {
    "location": {
      "column": 1,
      "line": 7,
    },
    "symbol": "}",
    "type": "symbol",
  },
]
`;
